crates/proto/src/udp/udp_client_stream.rs

```diff
@@ -291,73 +291,61 @@ async fn send_serial_message_inner<S: DnsUdpSocket + Send>(
             continue;
         }
 
-        match DnsResponse::from_buffer(response_buffer) {
-            Ok(response) => {
-                // Validate the message id in the response matches the value chosen for the query.
-                if msg_id != response.id() {
-                    // on wrong id, attempted poison?
-                    warn!(
-                        "expected message id: {} got: {}, dropped",
-                        msg_id,
-                        response.id()
-                    );
-
-                    continue;
-                }
+        let response = DnsResponse::from_buffer(response_buffer)?;
```

Comment:
This `?` will also drop unparseable messages, the difference is that this will return immediately with the decoding error, rather than returning later with a timeout error.

To gracefully handle messages with unparseable record data, we'd need a bigger change to how we represent messages internally, to allow holding uninterpreted bytes.

Reply:
> This ? will also drop unparseable messages, the difference is that this will return immediately with the decoding error, rather than returning later with a timeout error.

But that is an improvement, right?

Reply:
I think it might be a lateral move, and I don't think it accomplishes the title of the PR. In the case of DNSKEY records, we need to be able to pass through records with too-old or too-new algorithms to recursor clients, even if our crypto library doesn't support them. The main branch handles this by representing the public key as a byte buffer. Thus, we'll need to either keep that representation, add a fallback bytes-only representation to the `DNSKEY` struct if it primarily uses crypto library types, or store the original bytes higher up in `RData` or `Message`.

This also introduces a spoofing DoS vector: if an attacker floods a resolver or client with one-byte UDP packets with a spoofed source address and port, and destination ports in the 1024-65535 range, then they could trigger this error with a probability of min(1, 64512 / ((packets per second) * (latency of intended server))), causing the client to stop receiving before the real name server's reply arrives.

Reply:
> The main branch handles this by representing the public key as a byte buffer. Thus, we'll need to either keep that representation, add a fallback bytes-only representation to the `DNSKEY` struct if it primarily uses crypto library types, or store the original bytes higher up in `RData` or `Message`.

Yes, #2605 now has the `MaybePublicKey` enum for this as discussed in Discord, but it's not sufficient to pass the conformance tests yet. Do you even think the direction in #2605 is an improvement?